include/base_interface.h

```diff
@@ -0,0 +1,280 @@
+/**
+ * Base Interface
+ *
+ * Goal: provide a derivable interface to solve a particular
+ *       PDEs problem (time depending, first-order, non linear).
+ *
+ * Usage: This class requires some arguments related to the setting
+ *        of the problem: finite elements, boundary conditions,
+ *        and initial conditions.
+ *        Moreover, it helps to write the system matrix and
+ *        the preconditioner matrix.
+ * Variables:
+ *  - General:
+ *    - Finite Elements
+ *    - Boundary conditions ( Dirichlet, Neumann, and Robin )
+ *    - Initial conditions ( y(0) and d/dt y (0) )
+ *  - Couplings:
+ *    - coupling variable is a matrix of zeroes and ones: if the row
+ *      and the coloumn are indipendent there will be 0, otherwise 1
+ *  - @p default_differential_components : this variable is a list of ones and
+ *    zeroes. 1 in the case the corresponding variable should be differentiable
+ *    and 0 otherwise.
+ */
+
+#ifndef _pidomus_base_interface_h_
+#define _pidomus_base_interface_h_
+
+#include <deal.II/lac/linear_operator.h>
+/* #include <deal.II/lac/linear_operator.h> */
+/* #include <deal.II/lac/block_linear_operator.h> */
+/* #include <deal.II/lac/packaged_operation.h> */
+/* #include <deal.II/lac/solver_cg.h> */
+#include <deal.II/base/sacado_product_type.h>
+
+
+#include <deal2lkit/parsed_finite_element.h>
+#include <deal2lkit/parsed_function.h>
+#include <deal2lkit/parsed_mapped_functions.h>
+#include <deal2lkit/parsed_dirichlet_bcs.h>
+
+#include "copy_data.h"
+#include "lac/lac_type.h"
+
+using namespace pidomus;
+
+template <int dim,int spacedim=dim, typename LAC=LATrilinos>
+class BaseInterface : public ParameterAcceptor
+{
+
+public:
+
+  /**
+   * virtual destructor.
+   */
+  virtual ~BaseInterface() {};
+
+  /**
+   * Constructor. It takes the name of the subsection within the
+   * parameter file, the finite element used to discretize the system,
+   * the name of the components and a string were the block of
+   * differential and algebraic components are specified.
+   */
+  BaseInterface(const unsigned int &n_comp,
+                const unsigned int &n_matrices,
+                const std::string &name="",
+                const std::string &default_fe="FE_Q(1)",
+                const std::string &default_component_names="u",
+                const std::string &default_differential_components="");
+
+  void init ();
+
+
+  virtual void declare_parameters (ParameterHandler &prm);
+
+  virtual void parse_parameters_call_back() {};
+
+  /**
+   * returns the vector @p of differential blocks
```

Comment:
All functions documentation should be with an imperative form, following deal.II style: Return this. Make that. Do this.